### PR TITLE
feat(logger): improved logger

### PR DESCRIPTION
**Pull request by arturrez**

### CODE REVIEW

**Comment by arturrez** [comment]
```
{"level":"info","ts":"2022-08-15T22:06:16.663Z","logger":"controller-runtime.metrics","addr":"127.0.0.1:8080"}
{"level":"info","ts":"2022-08-15T22:06:16.664Z"}
{"level":"info","ts":"2022-08-15T22:06:16.994Z"}
{"level":"info","ts":"2022-08-15T22:06:17.007Z","milliseconds":343,"namespace":"avalanchego-operator-system","service":"avalanchego-operator-webhook-service","secret":"avalanchego-operator-webhook-server-cert"}
{"level":"info","ts":"2022-08-15T22:06:17.008Z","logger":"controller-runtime.builder","GVK":"chain.avax.network/v1alpha1, Kind=Avalanchego"}
{"level":"info","ts":"2022-08-15T22:06:17.008Z","logger":"controller-runtime.builder","GVK":"chain.avax.network/v1alpha1, Kind=Avalanchego","path":"/validate-chain-avax-network-v1alpha1-avalanchego"}
{"level":"info","ts":"2022-08-15T22:06:17.008Z","logger":"controller-runtime.webhook","path":"/validate-chain-avax-network-v1alpha1-avalanchego"}
{"level":"info","ts":"2022-08-15T22:06:17.008Z","logger":"controller-runtime.builder","GVK":"chain.avax.network/v1alpha1, Kind=AvalanchegoValidator"}
{"level":"info","ts":"2022-08-15T22:06:17.008Z","logger":"controller-runtime.builder","GVK":"chain.avax.network/v1alpha1, Kind=AvalanchegoValidator","path":"/validate-chain-avax-network-v1alpha1-avalanchegovalidator"}
{"level":"info","ts":"2022-08-15T22:06:17.008Z","logger":"controller-runtime.webhook","path":"/validate-chain-avax-network-v1alpha1-avalanchegovalidator"}
{"level":"info","ts":"2022-08-15T22:06:17.008Z","logger":"setup"}
I0815 22:06:17.009039       1 leaderelection.go:243] attempting to acquire leader lease avalanchego-operator-system/ecfcc342.avax.network...
{"level":"info","ts":"2022-08-15T22:06:17.009Z","path":"/metrics"}
{"level":"info","ts":"2022-08-15T22:06:17.009Z","logger":"controller-runtime.webhook.webhooks"}
{"level":"info","ts":"2022-08-15T22:06:17.010Z","logger":"controller-runtime.certwatcher"}
{"level":"info","ts":"2022-08-15T22:06:17.010Z","logger":"controller-runtime.webhook","host":"","port":9443}
{"level":"info","ts":"2022-08-15T22:06:17.048Z","logger":"controller-runtime.certwatcher"}
I0815 22:06:32.694966       1 leaderelection.go:253] successfully acquired lease avalanchego-operator-system/ecfcc342.avax.network
{"level":"info","ts":"2022-08-15T22:06:32.695Z","logger":"controller.avalanchegovalidator","reconciler group":"chain.avax.network","reconciler kind":"AvalanchegoValidator","source":"kind source: /, Kind="}
{"level":"info","ts":"2022-08-15T22:06:32.695Z","logger":"controller.avalanchegovalidator","reconciler group":"chain.avax.network","reconciler kind":"AvalanchegoValidator"}
{"level":"info","ts":"2022-08-15T22:06:32.695Z","logger":"controller.avalanchego","reconciler group":"chain.avax.network","reconciler kind":"Avalanchego","source":"kind source: /, Kind="}
{"level":"info","ts":"2022-08-15T22:06:32.695Z","logger":"controller.avalanchego","reconciler group":"chain.avax.network","reconciler kind":"Avalanchego"}
{"level":"info","ts":"2022-08-15T22:06:32.795Z","logger":"controller.avalanchego","reconciler group":"chain.avax.network","reconciler kind":"Avalanchego","worker count":1}
{"level":"info","ts":"2022-08-15T22:06:32.795Z","logger":"controller.avalanchegovalidator","reconciler group":"chain.avax.network","reconciler kind":"AvalanchegoValidator","worker count":1```
I am not sure why leader election is not picking up logger that we set in `manager` options

**Review by adutchak — APPROVED** [review]

File: controllers/avalanchego_controller.go [hunk 1]
```diff
@@ -68,7 +67,7 @@ type AvalanchegoReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/reconcile
 func (r *AvalanchegoReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	l := log.FromContext(ctx)
+	l := ctrl.LoggerFrom(ctx)
```
Comment (yevhenvolchenko) [inline controllers/avalanchego_controller.go:70]: can we have a single instance per Reconciler?
https://sdk.operatorframework.io/docs/building-operators/golang/references/logging/#creating-a-structured-log-statement

**Review by yevhenvolchenko — CHANGES_REQUESTED** [review]

File: controllers/utils/logger.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		LineEnding:     zapcore.DefaultLineEnding,
+		EncodeLevel:    zapcore.LowercaseLevelEncoder,
+		EncodeTime:     zapcore.ISO8601TimeEncoder,
+		EncodeDuration: zapcore.SecondsDurationEncoder,
+		EncodeCaller:   zapcore.ShortCallerEncoder,
+	}
+}
+
+// TextEncoderConfig returns an opinionated EncoderConfig
+func TextEncoderConfig() zapcore.EncoderConfig {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply controllers/utils/logger.go:27]: fixed

File: controllers/utils/logger.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+import (
+	"errors"
+
+	"go.uber.org/zap/zapcore"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+// JSONEncoderConfig returns an opinionated EncoderConfig
+func JSONEncoderConfig() zapcore.EncoderConfig {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply controllers/utils/logger.go:11]: fixed

File: main.go [hunk 4]
```diff
@@ -55,33 +59,46 @@ func main() {
...
 	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
 	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
 	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
 	flag.BoolVar(&enableWebhook, "webhook", true, "Enable validation webhook.")
-	opts := zap.Options{
-		Development: true,
-	}
-	opts.BindFlags(flag.CommandLine)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply main.go:67]: fixed

File: controllers/avalanchego_controller.go (hunk 1 above)
Comment (arturrez) [reply controllers/avalanchego_controller.go]: seems like this is what you suggest

**Review by yevhenvolchenko — APPROVED** [review]

**Pull request merged**